gadgets/src/algorithms/signature/schnorr_compressed.rs

```diff
@@ -0,0 +1,482 @@
+{
+    fn alloc_constant<
+        Fn: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<SchnorrCompressed<TE>>,
+        CS: ConstraintSystem<F>,
+    >(
+        _cs: CS,
+        value_gen: Fn,
+    ) -> Result<Self, SynthesisError> {
+        Ok(Self {
+            signature: (*value_gen()?.borrow()).clone(),
+            _engine: PhantomData,
+        })
+    }
+
+    fn alloc<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<SchnorrCompressed<TE>>, CS: ConstraintSystem<F>>(
+        _cs: CS,
+        _value_gen: Fn,
+    ) -> Result<Self, SynthesisError> {
+        unimplemented!()
+    }
+
+    fn alloc_input<
+        Fn: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<SchnorrCompressed<TE>>,
+        CS: ConstraintSystem<F>,
+    >(
+        _cs: CS,
+        _value_gen: Fn,
+    ) -> Result<Self, SynthesisError> {
+        unimplemented!()
+    }
+}
+
+impl<TE: TwistedEdwardsParameters<BaseField = F>, F: PrimeField + PoseidonDefaultParametersField>
+    SignatureGadget<SchnorrCompressed<TE>, F> for SchnorrCompressedGadget<TE, F>
+{
+    type PublicKeyGadget = SchnorrCompressedPublicKeyGadget<TE, F>;
+    type SignatureGadget = SchnorrCompressedSignatureGadget<TE, F>;
+
+    fn randomize_public_key<CS: ConstraintSystem<F>>(
+        &self,
+        mut cs: CS,
+        public_key: &Self::PublicKeyGadget,
+        randomizer: &[UInt8],
+    ) -> Result<Self::PublicKeyGadget, SynthesisError> {
+        let affine_zero: TEAffineGadget<TE, F> =
+            <TEAffineGadget<TE, F> as GroupGadget<TEAffine<TE>, F>>::zero(cs.ns(|| "affine zero")).unwrap();
+
+        let randomness = randomizer.iter().flat_map(|b| b.to_bits_le()).collect::<Vec<_>>();
+
+        let mut randomized_public_key = affine_zero;
+        randomized_public_key.scalar_multiplication(
+            cs.ns(|| "check_randomization_gadget"),
+            randomness.iter().zip_eq(&self.signature.generator_powers),
+        )?;
+        randomized_public_key = <TEAffineGadget<TE, F> as GroupGadget<TEAffine<TE>, F>>::add(
+            &randomized_public_key,
+            cs.ns(|| "pk + rG"),
+            &public_key.0,
+        )?;
+
+        Ok(SchnorrCompressedPublicKeyGadget {
+            0: randomized_public_key,
+        })
+    }
+
+    fn verify<CS: ConstraintSystem<F>>(
+        &self,
+        mut cs: CS,
+        public_key: &Self::PublicKeyGadget,
+        message: &[UInt8],
+        signature: &Self::SignatureGadget,
+    ) -> Result<Boolean, SynthesisError> {
+        let affine_zero: TEAffineGadget<TE, F> =
+            <TEAffineGadget<TE, F> as GroupGadget<TEAffine<TE>, F>>::zero(cs.ns(|| "affine zero")).unwrap();
+
+        let prover_response_bytes = signature
+            .prover_response
+            .to_bytes(cs.ns(|| "prover_response_to_bytes"))?;
+
+        let prover_response_bits = prover_response_bytes
+            .iter()
+            .flat_map(|byte| byte.to_bits_le())
+            .collect::<Vec<_>>();
```

Comment:
```suggestion
```
it should be possible to keep this iterator un-collected to avoid an allocation

Reply:
Yes, it works!